File: src/main/java/fr/anthonus/utils/api/RedditAPI.java
package fr.anthonus.utils.api;

import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import fr.anthonus.logs.LOGs;
import fr.anthonus.logs.logTypes.DefaultLogType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

import java.io.IOException;
import java.util.Random;

public class RedditAPI {
    private static final OkHttpClient client = new OkHttpClient();

    private static final String HEADER_NAME = "User-Agent";
    private static final String HEADER_VALUE = "ANTHONUS-Bot (https://github.com/ANTHONUSS/ANTHONUS-bot) by /u/Darkcp_YTB";

    public static String getBlursedImageURL(String subreddit) {
        String redditUrl = "https://www.reddit.com/r/" + subreddit + "/hot.json?limit=100";

        LOGs.sendLog("Création de la requête pour Reddit", DefaultLogType.API);
        Request request = new Request.Builder()
                .url(redditUrl)
                .header(HEADER_NAME, HEADER_VALUE)
                .build();

        LOGs.sendLog("Envoi de la requête à Reddit", DefaultLogType.API);
        String jsonData;
        try (Response response = client.newCall(request).execute()) {
            if (!response.isSuccessful()) {
                throw new IOException("Unexpected code " + response);
            }
            LOGs.sendLog("Requête recue", DefaultLogType.API);

            jsonData = response.body().string();

        } catch (IOException e) {
            throw new RuntimeException(e);
        }

        JsonArray posts = JsonParser.parseString(jsonData)
                .getAsJsonObject()
                .getAsJsonObject("data")
                .getAsJsonArray("children");

        JsonObject randomPost = getRandomPost(posts);
        while (randomPost == null) {
            randomPost = getRandomPost(posts);
        }

        return randomPost.get("url").getAsString();
    }

    private static JsonObject getRandomPost(JsonArray posts) {
        Random random = new Random();
        JsonObject randomPostData = posts.get(random.nextInt(posts.size()))
                .getAsJsonObject()
                .getAsJsonObject("data");

        if (randomPostData.has("url")) {
            String url = randomPostData.get("url").getAsString();
            if (url.endsWith(".jpg") || url.endsWith(".png") || url.endsWith(".gif")) {
                return randomPostData;
            }
        }

        return null;
    }

}

File: src/main/kotlin/api/RedditApi.kt
package api

import com.google.gson.JsonParser
import helpers.LogsHelper
import helpers.NetHelper
import okhttp3.OkHttpClient
import okhttp3.Request
import java.io.IOException

object RedditApi {
    private val client = NetHelper.client

    private const val HEADER_NAME = "user-agent"
    private const val HEADER_VALUE = "ANTHONUS-Bot (https://github.com/ANTHONUSS/ANTHONUS-bot) by /u/Darkcp_YTB"

    fun getRandomImage(subreddit: String): String? {
        val url = "https://www.reddit.com/r/$subreddit/new.json?limit=100"

        val request = Request.Builder()
            .header(HEADER_NAME, HEADER_VALUE)
            .url(url)
            .build()

        val jsonData = client.newCall(request).execute().use { response ->
            if (!response.isSuccessful) {
                LogsHelper.log.error("Response returned unexpected code", IOException("Unexpected code : $response"))
                return null
            }
            response.body.string()
        }

        try {
            val postsArray = JsonParser.parseString(jsonData)
                .asJsonObject
                .getAsJsonObject("data")
                .getAsJsonArray("children")

            val validImages = postsArray
                .map { it.asJsonObject.getAsJsonObject("data") }
                .filter { post ->
                    post.has("post_hint") && post.get("post_hint").asString == "image"
                            && post.has("is_video") && !post.get("is_video").asBoolean
                            && post.has("over_18") && !post.get("over_18").asBoolean
                            && post.has("url")
                }

            val randomPost = validImages.randomOrNull() ?: return null

            return randomPost.get("url").asString

        } catch (e: Exception) {
            LogsHelper.log.error("Error while parsing Json", e)
            return null
        }
    }
}